### ISSUE TITLE
Allow a trailing semicolon after the closing brace of a block in block-closing-brace-*-after

### ISSUE THREAD

**Issue by web-padawan** [stylelint/stylelint#1806]
CSS [custom property set](https://tabatkins.github.io/specs/css-apply-rule/#defining)  (used by `@apply` rules) is recognized as nested block due to its syntax, and so  triggers `block-closing-brace-newline-after` rule. 

This should be configurable by some extra option, e.g. `ignoreCustomPropertySets`.

> Which rule, if any, is this issue related to?

`block-closing-brace-newline-after`

> What CSS is needed to reproduce this issue?

```css
:root {
  --toolbar-theme: {
    background-color: hsl(120, 70%, 95%);
    border-radius: 4px;
    border: 1px solid var(--theme-color late);
  };
  --toolbar-title-theme: {
    color: green;
  };
}

.toolbar {
  @apply --toolbar-theme;
}
.toolbar > .title {
  @apply --toolbar-title-theme;
}

.warning {
  --toolbar-title-theme: {
    color: red;
    font-weight: bold;
  };
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "block-closing-brace-newline-before": "always-multi-line",
  }
}
```

> Which version of stylelint are you using?

`7.1.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, it's related to a [Tab Atkins draft](https://tabatkins.github.io/specs/css-apply-rule/)

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:

```shell
 6:6   ✖  Expected newline after "}"
```

**Comment by m-allanson** [comment]
I'd not heard of `@apply` before. Here's some info about it's current and possible future implementation status: https://www.chromestatus.com/feature/5753701012602880

**Comment by ntwb** [comment]
It would help if the CSS example actually included `@apply`, @web-padawan could you please update your example code.

**Comment by web-padawan** [comment]
@ntwb I updated the CSS example.
The usage of `@apply` is [shimmed](https://github.com/Polymer/polymer/blob/master/src/lib/apply-shim.html) by Polymer library, for the specific use case of [theming](https://www.polymer-project.org/1.0/docs/devguide/styling#custom-css-mixins) custom elements.

**Comment by m-allanson** [comment]
Thanks @web-padawan. I can see how this would be useful for working with Polymer. I think there's two approaches for this:

1. implement this functionality in a [plugin](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md)
1. add a new secondary option `ignoreSelector` (or similar) that accepts a string or regex pattern of selectors to ignore

Option 1 would probably be based on the existing `block-closing-brace-newline` rule, but with support for `@apply` syntax baked in.

While option 2 is 'just' adding an option to an existing rule, I think it's useful to look at the [guidelines for including a rule](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#let-everyone-know).

Given the above, I think this is a good candidate for a plugin (option 1), unless there are some more use-cases for option 2. 

Do stylelint/core have any thoughts on this?

**Comment by davidtheclark** [comment]
I'm not sure an option needs to be added to the existing rule ... we *could* just allow for a `;` before the expected newline, right? I think that would be simple and reasonable.

**Comment by web-padawan** [comment]
@davidtheclark That makes sense. In fact, a {}-wrapped block of properties is one value, and so must be followed by ';' (which by default can only be omitted for the last declaration).

**Comment by evilebottnawi** [comment]
@stylelint/core I think it is good reason create `stylelint-cssnext`, which have rules for newest stuff. After some time, some rules may include in core. We support all specs, include `draft`, right @davidtheclark ? But some newest stuff sometimes only in discussion, but many already use this. `stylelint-cssnext` repo be good solution.

**Comment by davidtheclark** [comment]
@evilebottnawi A `stylelint-cssnext` repo would be a solution for creating rules that are specifically about these `@apply` blocks. I totally agree. But the request here is about making the core rule not deliver false positives on `@apply` blocks, not about creating a new, `@apply`-specific rule. So I think we should treat it like other issues with SCSS or Less syntax: we just avoid messing up on it.

**Comment by evilebottnawi** [comment]
@davidtheclark agree

**Comment by jeddy3** [comment]
> we *could* just allow for a `;` before the expected newline, right?

Sgtm.

**Comment by web-padawan** [comment]
@jeddy3 One thing I'd like to say is that not only do we check for semicolon, but also for assignment.
Not any block should be allowed to have a `;`. See the difference:

This example below is valid: 
```css
:root {
  --x: { color: pink; };
}
```
And this one is not: 
```css
:root {
  --x { color: pink; };
}
```
Actually I think this should be implemented as a plugin.

**Comment by jeddy3** [comment]
> One thing I'd like to say is that not only do we check for semicolon

I think the plan is just ignore/remove any trailing semi-colons from after the closing brace *before* checking for the newline or space. Doing a simple patch like this doesn't seem to have a negative effect on the responsibility of this rule i.e. checking for spaces or newlines after the closing brace. I think that's the path forward to close this issue.

> Not any block should be allowed to have a `;`

We have a `no-extra-semicolons` rule, perhaps that should handle it. That's for another issue. Feel free to open.

> Actually I think this should be implemented as a plugin.

Yes, I think there's scope for plugins that checks for missing semicolons after custom property sets e.g. `stylelint-custom-property-set-no-missing-semicolon`, and one for missing colons before the open brace e.g. `stylelint-custom-property-set-no-missing-colon` . For another issue though.
